src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/IndexStateManagementIndices.kt

```diff
@@ -17,43 +17,97 @@ package com.amazon.opendistroforelasticsearch.indexstatemanagement
 import org.elasticsearch.action.ActionListener
+import org.elasticsearch.action.admin.indices.alias.Alias
 import org.elasticsearch.action.admin.indices.create.CreateIndexRequest
 import org.elasticsearch.action.admin.indices.create.CreateIndexResponse
+import org.elasticsearch.action.admin.indices.exists.indices.IndicesExistsRequest
+import org.elasticsearch.action.admin.indices.exists.indices.IndicesExistsResponse
+import org.elasticsearch.action.admin.indices.rollover.RolloverRequest
 import org.elasticsearch.client.Client
 import org.elasticsearch.client.IndicesAdminClient
+import org.elasticsearch.cluster.LocalNodeMasterListener
 import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.common.settings.Settings
 import org.elasticsearch.common.xcontent.XContentType
+import org.elasticsearch.threadpool.Scheduler.Cancellable
+import org.elasticsearch.threadpool.ThreadPool
 
 // TODO: Handle updating mappings on newer versions
-// TODO: Handle Auditing indices
 @OpenForTesting
 class IndexStateManagementIndices(
+    settings: Settings,
     private val client: IndicesAdminClient,
+    private val threadPool: ThreadPool,
     private val clusterService: ClusterService
-) {
+) : LocalNodeMasterListener {
 
+    private var scheduledRollover: Cancellable? = null
     private val logger = LogManager.getLogger(javaClass)
 
+    @Volatile private var historyMaxDocs = ISM_HISTORY_MAX_DOCS.get(settings)
+
+    @Volatile private var historyMaxAge = ISM_HISTORY_INDEX_MAX_AGE.get(settings)
+
+    @Volatile private var historyRolloverCheckPeriod = ISM_HISTORY_ROLLOVER_CHECK_PERIOD.get(settings)
+
+    private val indexStateManagementMappings = javaClass.classLoader.getResource("mappings/opendistro-ism-config.json").readText()
+
+    private val indexStateManagementHistoryMappings = javaClass.classLoader.getResource("mappings/opendistro-ism-history.json").readText()
+
+    init {
+        clusterService.addLocalNodeMasterListener(this)
+        clusterService.clusterSettings.addSettingsUpdateConsumer(ISM_HISTORY_MAX_DOCS) { historyMaxDocs = it }
+        clusterService.clusterSettings.addSettingsUpdateConsumer(ISM_HISTORY_INDEX_MAX_AGE) { historyMaxAge = it }
+        clusterService.clusterSettings.addSettingsUpdateConsumer(ISM_HISTORY_ROLLOVER_CHECK_PERIOD) {
+            historyRolloverCheckPeriod = it
+            rescheduleRollover()
+        }
+    }
+
+    override fun onMaster() {
+        try {
+            // try to rollover immediately as we might be restarting the cluster
+            rolloverHistoryIndex()
+            // schedule the next rollover for approx MAX_AGE later
+            scheduledRollover = threadPool.scheduleWithFixedDelay({ rolloverHistoryIndex() }, historyRolloverCheckPeriod, executorName())
+        } catch (e: Exception) {
+            // This should be run on cluster startup
+            logger.error("Error creating ISM history index. History can't be recorded until master node is restarted.", e)
```

Comment:
Just for clarification, based on the comment above saying ```This should be run on cluster startup```, is this log message saying that the master node needs to be restarted (implying manual intervention) or that we need to wait for the master node to come up since the cluster is starting up?

Reply:
This really isn't true. Updated the error message.